libs/utils/include/utils/vector.h

```diff
@@ -55,6 +55,214 @@ static inline bool insert_sorted_unique(std::vector<T>& v, T item) {
     return false;
 }
 
+//  -----------------------------------------------------------------------------------------------
+
+class TrivialVectorBase {
+protected:
+    using size_type = uint32_t;
+
+    char* mBegin = nullptr;
+    size_type mItemCount = 0;
+    size_type mCapacity = 0;
+
+    TrivialVectorBase() noexcept = default;
+
+    TrivialVectorBase(size_type itemSize, size_type count) noexcept
+            : mItemCount(count), mCapacity(count) {
+        mBegin = (char*)malloc(mCapacity * itemSize);  // FIXME: move this in cpp
+    }
+
+    TrivialVectorBase(TrivialVectorBase&& rhs) noexcept {
+        this->swap(rhs);
+    }
+
+    ~TrivialVectorBase() noexcept {
+        free(mBegin);
+    }
+
+    inline void assert_capacity(size_type c, size_type itemSize) {
+        if (UTILS_UNLIKELY(mCapacity < c)) {
+            assert_capacity_slow(c, itemSize);
+        }
+    }
+
+    UTILS_NOINLINE
+    void assert_capacity_slow(size_type c, size_type itemSize) {
+        c = (c * 3 + 1) / 2;
+        set_capacity(c, itemSize);
+    }
+
+    UTILS_NOINLINE
+    void set_capacity(size_type n, size_type itemSize) {
+        // FIXME: move this in cpp
+        if (n == mCapacity) {
+            return;
+        }
+        char* addr = (char*)malloc(n * itemSize);
+        memcpy(addr, mBegin, ((mItemCount < n) ? mItemCount : n) * itemSize);
+        free(mBegin);
+        mBegin = addr;
+        mCapacity = n;
+    }
+
+    void swap(TrivialVectorBase& other) {
+        using std::swap;
+        swap(mBegin, other.mBegin);
+        swap(mItemCount, other.mItemCount);
+        swap(mCapacity, other.mCapacity);
+    }
+
+    char* begin() noexcept { return mBegin; }
+    char* end(size_type itemSize) noexcept { return mBegin + mItemCount * itemSize; }
+    char const* begin() const noexcept { return const_cast<TrivialVectorBase*>(this)->begin(); }
+    char const* end(size_type itemSize) const noexcept { return const_cast<TrivialVectorBase*>(this)->end(itemSize); }
+};
+
+template<typename T, std::enable_if_t<std::is_trivial<T>::value, bool> = true>
+class vector : private TrivialVectorBase {
```

Comment:
Should we give it a different name to avoid using the wrong one/collisions when/if we do using utils and using std in the same file? I think we rarely if ever do using std but…